src/datasources/accounts/address-books/address-books.datasource.ts

```diff
@@ -0,0 +1,124 @@
+  }
+
+  async getAddressBook(account: Account): Promise<AddressBook> {
+    const [dbAddressBook] = await this.sql<
+      DbAddressBook[]
+    >`SELECT * FROM address_books WHERE account_id = ${account.id}`;
+    if (!dbAddressBook) throw new AddressBookNotFoundError();
+    return this.addressBookMapper.map(dbAddressBook);
+  }
+
+  async updateAddressBookItem(args: {
+    addressBook: AddressBook;
+    updateAddressBookItem: UpdateAddressBookItemDto;
+  }): Promise<AddressBookItem> {
+    const { addressBook, updateAddressBookItem } = args;
+    const addressBookItem = addressBook.data.find(
+      (i) => i.id === updateAddressBookItem.id,
+    );
+    if (!addressBookItem) throw new AddressBookItemNotFoundError();
+    addressBookItem.address = updateAddressBookItem.address;
+    addressBookItem.name = updateAddressBookItem.name;
+    await this.updateAddressBook(addressBook);
+    return addressBookItem;
+  }
+
+  async deleteAddressBook(addressBook: AddressBook): Promise<void> {
+    await this.sql`DELETE FROM address_books WHERE id = ${addressBook.id}`;
+  }
+
+  async deleteAddressBookItem(args: {
+    addressBook: AddressBook;
+    id: number;
+  }): Promise<void> {
+    const { addressBook, id } = args;
+    const index = addressBook.data.findIndex((i) => i.id === id);
+    if (index === -1) throw new AddressBookItemNotFoundError();
+    addressBook.data.splice(index, 1);
+    await this.updateAddressBook(addressBook);
+  }
+
+  private async getOrCreateAddressBook(account: Account): Promise<AddressBook> {
+    try {
+      return await this.getAddressBook(account);
+    } catch (err) {
+      if (err instanceof AddressBookNotFoundError) {
+        return this.addressBookMapper.map(
+          await this.createAddressBook({ account }),
+        );
+      }
+      throw err;
+    }
+  }
+
+  private async createAddressBook(args: {
+    account: Account;
+  }): Promise<DbAddressBook> {
+    const encryptionApi = await this.encryptionApiManager.getApi();
+    const encryptedBlob = await encryptionApi.encryptBlob([]);
+    const [dbAddressBook] = await this.sql<DbAddressBook[]>`
+      INSERT INTO address_books (data, key, iv, account_id)
+      VALUES (
+        ${encryptedBlob.encryptedData},
+        ${encryptedBlob.encryptedDataKey},
+        ${encryptedBlob.iv},
+        ${args.account.id}
+      ) RETURNING *;
+    `;
+    return dbAddressBook;
+  }
+
+  private async updateAddressBook(
+    addressBook: AddressBook,
+  ): Promise<DbAddressBook> {
+    const encryptionApi = await this.encryptionApiManager.getApi();
+    const encryptedBlob = await encryptionApi.encryptBlob(addressBook.data);
+    const [dbAddressBook] = await this.sql<DbAddressBook[]>`
+      UPDATE address_books
+      SET data = ${encryptedBlob.encryptedData},
+          key = ${encryptedBlob.encryptedDataKey},
+          iv = ${encryptedBlob.iv}
+      WHERE id = ${addressBook.id}
+     RETURNING *;
+    `;
+    return dbAddressBook;
+  }
```

Comment:
So essentially it overwrites the whole address book each time because it has to be re-encrypted? Sort of defeats the purpose of SQL but I guess it's better to have one database for everything than several. :)